File: 0226_Invert_Binary_Tree/cpp/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
// Recursive T = O(n), S = O(n)
class Solution {
  public:
    TreeNode *invertTree(TreeNode *root) {
        if (root == nullptr) {
            return nullptr;
        }
        swap(root->left, root->right);
        invertTree(root->left);
        invertTree(root->right);
        return root;
    }
};

// Iteration T = O(n), S = O(n)
class Solution {
  public:
    TreeNode *invertTree(TreeNode *root) {
        if (root == nullptr) {
            return nullptr;
        }
        queue<TreeNode *> q;
        q.push(root);
        while (!q.empty()) {
            TreeNode *p = q.front();
            q.pop();
            swap(p->left, p->right);
            if (p->left) {
                q.push(p->left);
            }
            if (p->right) {
                q.push(p->right);
            }
        }
        return root;
    }
};

File: 0226_Invert_Binary_Tree/golang/main.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// Recursive T = O(n), S = O(n)
func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	dfs(root)
	return root
}

func dfs(root *TreeNode) {
	if root == nil {
		return
	}
	root.Left, root.Right = root.Right, root.Left
	dfs(root.Left)
	dfs(root.Right)
}

// Iteration T = O(n), S = O(n)
func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	st := []*TreeNode{root}
	for len(st) != 0 {
		p := st[len(st)-1]
		st = st[0 : len(st)-1]
		p.Left, p.Right = p.Right, p.Left
		if p.Left != nil {
			st = append(st, p.Left)
		}
		if p.Right != nil {
			st = append(st, p.Right)
		}
	}
	return root
}